### PR TITLE
fix panic related to commit 41d14568

### PR DESCRIPTION
**Pull request by basharal**
We shouldn't write to the channel after it's closed upon termination

### CODE REVIEW

File: marshal/claim.go [hunk 1]
```diff
@@ -229,32 +239,18 @@ func (c *claim) GetCurrentLag() int64 {
 	return 0
 }
 
-// Release will invoke our release mechanism if and only if we are still claimed.
-func (c *claim) Release() bool {
-	if !atomic.CompareAndSwapInt32(c.claimed, 1, 0) {
+// Release will invoke commit offsets and may release the Kafka partition depending on whether
+// releasePartition was passed or not
+func (c *claim) Release(releasePartition bool) bool {
```
Comment (zorkian) [inline marshal/claim.go]: I think the interface to Release should be that it releases the partition. I'm fine with having a Terminate method that can be used to release or not release, but I think a method called Release should always do what it says on the tin.

Comment (basharal) [inline marshal/claim.go]: Sounds good...I just wanted to share the implementation...I will create two separate functions...

File: marshal/claim.go [hunk 2]
```diff
@@ -311,7 +314,16 @@ func (c *claim) messagePump() {
 		c.lock.Unlock()
 
 		// Push the message down to the client (this bypasses the Consumer)
-		c.messages <- msg
+		// We should NOT write to the consumer channel if the claim is no longer claimed. This
+		// needs to be serialized with Release, otherwise a race-condition can potentially
+		// lead to a write to a closed-channel. That's why we're using this lock. We're not
+		// using the main lock to avoid deadlocks since the write to the channel is blocking
+		// until someone consumes the message blocking all Commit operations
+		c.messagesLock.Lock()
```
Comment (zorkian) [inline marshal/claim.go]: I'm pretty dubious about adding a second lock here. I'm trying to think of a way to handle this without locking, but nothing has immediately come to mind... I will think about it some more over the weekend and try to come up with something

Comment (basharal) [inline marshal/claim.go]: I thought about this and couldn't think of a better way of solving this...the main problem is that the creator of the channel (the consumer) is not the one writing to it....this is the current design....
I'll be happy to change this if you managed to think of something over the weekend

Comment (zorkian) [inline marshal/claim.go]: I didn't, so we'll take this

**Pull request merged**